finagle-core/src/main/scala/com/twitter/finagle/Filter.scala

```diff
@@ -126,6 +109,49 @@ abstract class Filter[-ReqIn, +RepOut, +ReqOut, -RepIn]
 abstract class SimpleFilter[Req, Rep] extends Filter[Req, Rep, Req, Rep]
 
 object Filter {
+  private case class AndThen[ReqIn, RepOut, ReqOut, RepIn](
+      build: Service[ReqOut, RepIn] => Service[ReqIn, RepOut]
+  )
+    extends Filter[ReqIn, RepOut, ReqOut, RepIn]
+  {
+    override def andThen[Req2, Rep2](
+      next: Filter[ReqOut, RepIn, Req2, Rep2]
+    ): Filter[ReqIn, RepOut, Req2, Rep2] =
+      AndThen(service => build(next.andThen(service)))
+
+    override def andThen(service: Service[ReqOut, RepIn]): Service[ReqIn, RepOut] =
+      build(service)
+
+    override def andThen(
+      factory: ServiceFactory[ReqOut, RepIn]
+    ): ServiceFactory[ReqIn, RepOut] =
+      new ServiceFactory[ReqIn, RepOut] {
```

Comment:
we can skip some of the boilerplate with `ServiceFactoryProxy`, and then override apply.